stdlib/public/runtime/Paths.cpp

```diff
@@ -0,0 +1,369 @@
+///
+///  https://learn.microsoft.com/en-us/windows/win32/memory/obtaining-a-file-name-from-a-file-handle
+///
+/// @param pszFilename The NT-style filename to convert.
+///
+/// @result A string, allocated using std::malloc(), containing the Win32-style
+///         filename.
+LPWSTR
+_swift_win32NameFromNTName(LPWSTR pszFilename) {
+  DWORD dwLen = GetLogicalDriveStringsW(0, NULL);
+  if (!dwLen)
+    return NULL;
+
+  LPWSTR lpDriveStrings = (LPWSTR)std::malloc(dwLen * sizeof(WCHAR));
+  if (!lpDriveStrings)
+    return NULL;
+
+  DWORD dwRet = GetLogicalDriveStringsW(dwLen, lpDriveStrings);
+  if (!dwRet)
+    return NULL;
+
+  LPWSTR pszDrive = lpDriveStrings;
+  while (*pszDrive) {
+    size_t len = wcslen(pszDrive);
+    if (len && pszDrive[len - 1] == '\\')
+      pszDrive[len - 1] = 0;
+
+    WCHAR ntPath[4096];
+    dwRet = QueryDosDeviceW(pszDrive, ntPath, 4096);
+    if (dwRet) {
+      size_t ntLen = wcslen(ntPath);
+
+      if (_wcsnicmp(pszFilename, ntPath, ntLen) == 0
+          && pszFilename[ntLen] == '\\') {
+        size_t fnLen = wcslen(pszFilename);
+        size_t driveLen = wcslen(pszDrive);
+        size_t pathLen = fnLen - ntLen;
+        size_t newLen = driveLen + pathLen + 1;
+        LPWSTR pszWin32Name = (LPWSTR)std::malloc(newLen * sizeof(WCHAR));
+        if (!pszWin32Name) {
+          std::free(lpDriveStrings);
+          return NULL;
+        }
+
+        LPWSTR ptr = pszWin32Name;
+        memcpy(ptr, pszDrive, driveLen * sizeof(WCHAR));
+        ptr += driveLen;
+        memcpy(ptr, pszFilename + ntLen, pathLen * sizeof(WCHAR));
+        ptr += pathLen;
+        *ptr = 0;
+
+        std::free(lpDriveStrings);
+
+        return pszWin32Name;
+      }
+    }
+
+    pszDrive += len + 1;
+  }
+
+  std::free(lpDriveStrings);
+
+  return _wcsdup(pszFilename);
+}
+#endif
+
+}
+
+SWIFT_RUNTIME_EXPORT
+const char *
+swift_getAuxiliaryExecutablePath(const char *name)
+{
+  const char *rootPath = swift_getRootPath();
+
+  // Form <rootPath>/libexec/swift/
+  size_t rootPathLen = std::strlen(rootPath);
+  const char *libexecStr = PATHSEP_STR "libexec" PATHSEP_STR "swift" PATHSEP_STR;
+  size_t libexecLen = std::strlen(libexecStr);
+  char *libexecPath = (char *)malloc(rootPathLen + libexecLen + 1);
+  std::memcpy(libexecPath, rootPath, rootPathLen);
+  std::memcpy(libexecPath + rootPathLen, libexecStr, libexecLen + 1);
+
+  // If libexec exists, look there
+  if (_swift_exists(libexecPath)) {
+    const char *result = _swift_getAuxExePathIn(libexecPath, name);
+
+    free(libexecPath);
+
+    return result;
+  }
+
+  free(libexecPath);
+
+  // Otherwise, look in the root itself
+  return _swift_getAuxExePathIn(rootPath, name);
+}
+
+namespace {
+
+const char *
+_swift_getAuxExePathIn(const char *path, const char *name)
+{
+  size_t pathLen = std::strlen(path);
+  size_t nameLen = std::strlen(name);
+  size_t extLen = 0;
+
+#ifdef _WIN32
+  if (nameLen > 4 && strcmp(name + nameLen - 4, ".exe") != 0)
```

Comment:
Does it make more sense to use `PathCchFindExtension`?  I worry about silly things like XCTest which uses `.xctest` on Windows (yes, really).

Reply:
I don't think it'll be a problem in this instance; the point of this was to let us write the name without the extension and have the Windows code add `.exe` automatically. I could have done it unconditionally, but I figured there was always a chance someone would pass something with `.exe` to it.

I don't think we want this except for executables (i.e. I don't think we should be using it for `.dll`s, because then it'll get much more complicated as they're `.so`s on Linux, `.dylib`s on macOS and potentially have other extensions on other platforms too).